mountpoint-s3/src/inode/negative_cache.rs

```diff
@@ -0,0 +1,204 @@
+                    e.is_valid().into()
+                );
+            }
+            metrics::gauge!("metadata_cache.negative_cache.entries", map.len() as f64);
+        }
+
+        metrics::histogram!(
+            "metadata_cache.negative_cache.operation_duration_us",
+            start.elapsed().as_micros() as f64,
+            "op" => "insert",
+        );
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{thread::sleep, time::Duration};
+
+    use super::NegativeCache;
+
+    #[test]
+    fn test_contains() {
+        let cache = NegativeCache::new(100, Duration::from_secs(60));
+
+        cache.insert(1, "child1");
+        assert!(cache.contains(1, "child1"));
+        assert!(!cache.contains(1, "child2"));
+        assert!(!cache.contains(2, "child1"));
+    }
+
+    #[test]
+    fn test_insert() {
+        let cache = NegativeCache::new(100, Duration::from_secs(60));
+
+        cache.insert(1, "child1");
+        assert!(cache.contains(1, "child1"));
+
+        cache.insert(1, "child2");
+        assert!(cache.contains(1, "child2"));
+        assert!(cache.contains(1, "child1"));
+
+        cache.insert(2, "child1");
+        assert!(cache.contains(2, "child1"));
+        assert!(cache.contains(1, "child2"));
+        assert!(cache.contains(1, "child1"));
+    }
+
+    #[test]
+    fn test_remove() {
+        let cache = NegativeCache::new(100, Duration::from_secs(60));
+
+        cache.insert(1, "child1");
+        cache.insert(1, "child2");
+        cache.insert(2, "child1");
+        assert!(cache.contains(1, "child1"));
+        assert!(cache.contains(1, "child2"));
+        assert!(cache.contains(2, "child1"));
+
+        cache.remove(1, "child1");
+        assert!(!cache.contains(1, "child1"));
+        assert!(cache.contains(1, "child2"));
+        assert!(cache.contains(2, "child1"));
+    }
+
+    #[test]
+    fn test_max_size() {
+        let cache = NegativeCache::new(2, Duration::from_secs(60));
+
+        cache.insert(1, "child1");
+        assert!(cache.contains(1, "child1"));
+
+        cache.insert(1, "child2");
+        assert!(cache.contains(1, "child2"));
+        assert!(cache.contains(1, "child1"));
+
+        cache.insert(1, "child3");
+        assert!(cache.contains(1, "child3"));
+        assert!(cache.contains(1, "child2"));
+        assert!(!cache.contains(1, "child1"));
+    }
+
+    #[test]
+    fn test_expiration() {
+        let cache = NegativeCache::new(100, Duration::from_millis(1));
+
+        cache.insert(1, "child1");
+        sleep(Duration::from_millis(2));
+        assert!(!cache.contains(1, "child1"));
+    }
+
+    #[test]
+    fn test_insert_resets_validity() {
+        let cache = NegativeCache::new(100, Duration::from_millis(50));
+
+        cache.insert(1, "child1");
+        sleep(Duration::from_millis(100));
+        assert!(!cache.contains(1, "child1"));
+
+        cache.insert(1, "child1");
+        assert!(cache.contains(1, "child1"));
+    }
```

Comment:
This test doesn't quite test the validity reset. I think we want something more like:

```suggestion
    #[test]
    fn test_insert_resets_validity() {
        let cache = NegativeCache::new(100, Duration::from_millis(100));

        cache.insert(1, "child1");
        sleep(Duration::from_millis(50));
        cache.insert(1, "child1"); // refresh
        sleep(Duration::from_millis(75)); // original 100ms exceeded
        assert!(cache.contains(1, "child1"));
    }
```